mpclient_x64.c

```diff
@@ -0,0 +1,260 @@
+#include <err.h>
+
+#include "winnt_types.h"
+#include "pe_linker.h"
+#include "ntoskernel.h"
+#include "util.h"
+#include "log.h"
+#include "rsignal.h"
+#include "engineboot.h"
+#include "scanreply.h"
+#include "streambuffer.h"
+#include "openscan.h"
+#include "hook.h"
+#include "mpclient.h"
+
+struct pe_image image = {
+        .entry  = NULL,
+        .name   = "engine/x64/mpengine.dll",
+};
+
+// Any usage limits to prevent bugs disrupting system.
+const struct rlimit kUsageLimits[] = {
+    [RLIMIT_FSIZE]  = { .rlim_cur = 0x20000000, .rlim_max = 0x20000000 },
+    [RLIMIT_CPU]    = { .rlim_cur = 3600,       .rlim_max = RLIM_INFINITY },
+    [RLIMIT_CORE]   = { .rlim_cur = 0,          .rlim_max = 0 },
+    [RLIMIT_NOFILE] = { .rlim_cur = 32,         .rlim_max = 32 },
+};
+
+DWORD WINAPI (* __rsignal)(PHANDLE KernelHandle, DWORD Code, PVOID Params, DWORD Size);
+
+static DWORD WINAPI EngineScanCallback(PSCANSTRUCT Scan)
+{
+    if (Scan->Flags & SCAN_MEMBERNAME) {
+        LogMessage("Scanning archive member %s", Scan->VirusName);
+    }
+    if (Scan->Flags & SCAN_FILENAME) {
+        LogMessage("Scanning %s", Scan->FileName);
+    }
+    if (Scan->Flags & SCAN_PACKERSTART) {
+        LogMessage("Packer %s identified.", Scan->VirusName);
+    }
+    if (Scan->Flags & SCAN_ENCRYPTED) {
+        LogMessage("File is encrypted.");
+    }
+    if (Scan->Flags & SCAN_CORRUPT) {
+        LogMessage("File may be corrupt.");
+    }
+    if (Scan->Flags & SCAN_FILETYPE) {
+        LogMessage("File %s is identified as %s", Scan->FileName, Scan->VirusName);
+    }
+    if (Scan->Flags & 0x08000022) {
+        LogMessage("Threat %s identified.", Scan->VirusName);
+    }
+    // This may indicate PUA.
+    if ((Scan->Flags & 0x40010000) == 0x40010000) {
+        LogMessage("Threat %s identified.", Scan->VirusName);
+    }
+    return 0;
+}
+
+static DWORD WINAPI ReadStream(PVOID this, ULONGLONG Offset, PVOID Buffer, DWORD Size, PDWORD SizeRead)
+{
+    fseek(this, Offset, SEEK_SET);
+    *SizeRead = fread(Buffer, 1, Size, this);
+    return TRUE;
+}
+
+static DWORD WINAPI GetStreamSize(PVOID this, PULONGLONG FileSize)
+{
+    fseek(this, 0, SEEK_END);
+    *FileSize = ftell(this);
+    return TRUE;
+}
+
+static PWCHAR WINAPI GetStreamName(PVOID this)
+{
+    return L"input";
+}
+
+// These are available for pintool.
+BOOL __noinline InstrumentationCallback(PVOID ImageStart, SIZE_T ImageSize)
+{
+    // Prevent the call from being optimized away.
+    asm volatile ("");
+    return TRUE;
+}
+
+int main(int argc, char **argv, char **envp)
+{
+    PIMAGE_DOS_HEADER DosHeader;
+    PIMAGE_NT_HEADERS PeHeader;
+    HANDLE KernelHandle;
+    SCAN_REPLY ScanReply;
+    BOOTENGINE_PARAMS BootParams;
+    SCANSTREAM_PARAMS ScanParams;
+    STREAMBUFFER_DESCRIPTOR ScanDescriptor;
+    ENGINE_INFO EngineInfo;
+    ENGINE_CONFIG EngineConfig;
+
+    // Load the mpengine module.
+    if (pe_load_library(image.name, &image.image, &image.size) == false) {
```

Comment:
I think there's a bug: `image.size` is an `int` but `pe_load_library` takes a `size_t*`.